File: tests/wpt/web-platform-tests/html/cross-origin-embedder-policy/credentialless/resources/dispatcher.py
import json
from wptserve.utils import isomorphic_decode

# A server used to store and retrieve arbitrary data.
# This is used by: ./dispatcher.js
def main(request, response):
    # This server is configured so that is accept to receive any requests and
    # any cookies the web browser is willing to send.
    response.headers.set(b"Access-Control-Allow-Credentials", b"true")
    response.headers.set(b'Access-Control-Allow-Methods', b'OPTIONS, GET, POST')
    response.headers.set(b'Access-Control-Allow-Headers', b'Content-Type')
    response.headers.set(b'Cache-Control', b'no-cache, no-store, must-revalidate')
    response.headers.set(b"Access-Control-Allow-Origin", request.headers.get(b"origin") or '*')

    # CORS preflight
    if request.method == u'OPTIONS':
        return b''

    uuid = request.GET[b'uuid']
    stash = request.server.stash;

    # The stash is accessed concurrently by many clients. A lock is used to
    # avoid unterleaved read/write from different clients.
    with stash.lock:
        queue = stash.take(uuid) or [];

        # Push into the |uuid| queue, the requested headers.
        if b"show-headers" in request.GET:
            headers = {};
            for key, value in request.headers.items():
                headers[isomorphic_decode(key)] = isomorphic_decode(request.headers[key])
            headers = json.dumps(headers);
            queue.append(headers);
            ret = headers;

        # Push into the |uuid| queue, the posted data.
        elif request.method == u'POST':
            queue.append(request.body)
            ret = b'done'

        # Pull from the |uuid| queue, the posted data.
        else:
            if len(queue) == 0:
                ret = b'not ready'
            else:
                ret = queue.pop(0)

        stash.put(uuid, queue)
    return ret;

File: tests/wpt/web-platform-tests/html/cross-origin-embedder-policy/credentialless/resources/dispatcher.js
// Define an universal message passing API. It works cross-origin and across
// browsing context groups.
const dispatcher_path =
  "/html/cross-origin-embedder-policy/credentialless/resources/dispatcher.py";
const dispatcher_url = new URL(dispatcher_path, location.href).href;

// Return a promise, limiting the number of concurrent accesses to a shared
// resources to |max_concurrent_access|.
const concurrencyLimiter = (max_concurrency) => {
  let pending = 0;
  let waiting = [];
  return async (task) => {
    pending++;
    if (pending > max_concurrency)
      await new Promise(resolve => waiting.push(resolve));
    let result = await task();
    pending--;
    waiting.shift()?.();
    return result;
  };
}

// Wait for a random amount of time in the range [10ms,100ms].
const randomDelay = () => {
  return new Promise(resolve => setTimeout(resolve, 10 + 90*Math.random()));
}

// Sending too many requests in parallel causes congestion. Limiting it improves
// throughput.
//
// Note: The following table has been determined on the test:
// ../cache-storage.tentative.https.html
// using Chrome with a 64 core CPU / 64GB ram, in release mode:
// ┌───────────┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬────┐
// │concurrency│ 1 │ 2 │ 3 │ 4 │ 5 │ 6 │ 10│ 15│ 20│ 30│ 50│ 100│
// ├───────────┼───┼───┼───┼───┼───┼───┼───┼───┼───┼───┼───┼────┤
// │time (s)   │ 54│ 38│ 31│ 29│ 26│ 24│ 22│ 22│ 22│ 22│ 34│ 36 │
// └───────────┴───┴───┴───┴───┴───┴───┴───┴───┴───┴───┴───┴────┘
const limiter = concurrencyLimiter(6);

const send = async function(uuid, message) {
  await limiter(async () => {
    // Requests might be dropped. Retry until getting a confirmation it has been
    // processed.
    while(1) {
      try {
        let response = await fetch(dispatcher_url + `?uuid=${uuid}`, {
          method: 'POST',
          body: message
        })
        if (await response.text() == "done")
          return;
      } catch (fetch_error) {}
      await randomDelay();
    };
  });
}

const receive = async function(uuid) {
  while(1) {
    let data = "not ready";
    try {
      data = await limiter(async () => {
        let response = await fetch(dispatcher_url + `?uuid=${uuid}`);
        return await response.text();
      });
    } catch (fetch_error) {}

    if (data == "not ready") {
      await randomDelay();
      continue;
    }

    return data;
  }
}

// Returns an URL. When called, the server sends toward the `uuid` queue the
// request headers. Useful for determining if something was requested with
// Cookies.
const showRequestHeaders= function(origin, uuid) {
  return origin + dispatcher_path + `?uuid=${uuid}&show-headers`;
}
